### PR TITLE
feat: all unverified inbound sessions

### PR DESCRIPTION
**Pull request by acolytec3**
Features
- Adds new optional `sessionConfig` parameter to allow inbound sessions with invalid ENRs (i.e. where observed socket address doesn't match socket address reported in ENR - to allow nodes reporting local IP/port combinations to receive `discv5` PONG messages in order to update ENRs).
  - Note: nodes with "unverified sessions" are not added to the `discv5` routing table so the discv5 service will never send outbound messages to them to ensure we aren't propagating DDoS attacks
- Allows `sendTalkResp` to specify a target node via session key (`nodeId + socketAddr`) to allow responding to nodes with unverified sessions (so will not have an ENR in the routing table)
- Makes `service.sendPing` public so it can be triggered by portal network layer to trigger `enrUpdate` logic via PONG messages.

Fixes
- Fixes tests

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -13,7 +13,7 @@
     "lint": "eslint --color --ext .ts src/",
     "test": "yarn test:unit && yarn test:e2e",
     "test:unit": "mocha 'test/unit/**/*.test.ts'",
-    "test:e2e": "mocha 'test/unit/**/*.test.ts'"
+    "test:e2e": "mocha 'test/e2e/**/*.test.ts'"
```
Comment (dapplion) [inline package.json:16]: :pray:

**Comment by acolytec3** [comment]
This should be ready for review.

File: src/service/service.ts [hunk 2]
```diff
@@ -386,7 +386,11 @@ export class Discv5 extends (EventEmitter as { new (): Discv5EventEmitter }) {
   /**
    * Send TALKRESP message to requesting node
    */
-  public async sendTalkResp(remote: ENR | Multiaddr, requestId: RequestId, payload: Uint8Array): Promise<void> {
+  public async sendTalkResp(
+    remote: ENR | Multiaddr | INodeAddress,
```
Comment (wemeetagain) [inline src/service/service.ts:390]: Can you change this to only allow `nodeAddr: INodeAddress`?

And revert the changes to `createNodeContact`?

File: src/service/service.ts [hunk 3]
```diff
@@ -402,7 +406,7 @@ export class Discv5 extends (EventEmitter as { new (): Discv5EventEmitter }) {
   /**
    * Sends a PING request to a node
    */
-  private sendPing(enr: ENR): void {
+  public sendPing(enr: ENR): void {
```
Comment (wemeetagain) [inline src/service/service.ts:409]: Can you change the param to `remote: ENR | Multiaddr`? Would allow pinging a peer via full multiaddr.

**Review by wemeetagain — CHANGES_REQUESTED** [review]
Does it make sense to also also allow outbound unverified sessions?
They seem less useful, but you could dial a peer via a full multiaddr, and currently, if they respond with 'unverified' ENR (wrong or no IP+port), the session is dropped.
If we allowed unverified outbound sessions, we'd be able to use them for TALKREQ/RESP at least.

File: src/service/service.ts (hunk 2 above)
Comment (acolytec3) [reply src/service/service.ts:390]: To make sure I'm tracking, you want `remote : INodeAddress`?  If so, don't we need to update the function signature of `sendRpcResponse`  as well since it wants a `NodeContact` and not just the `INodeAddress`.  The challenge this gives for Portal Network (unless I'm missing something) is that we may not have access to a node's public key, just an invalid ENR that contains the nodeId.  We can use this `nodeId` plus the observed address to still specify which session to use but there's no clear way to get the full Multiaddr with an inlined public key from that since as far as I understand you can't go from a nodeId back to a public key.

**Comment by acolytec3** [comment]
> Does it make sense to also also allow outbound unverified sessions? They seem less useful, but you could dial a peer via a full multiaddr, and currently, if they respond with 'unverified' ENR (wrong or no IP+port), the session is dropped. If we allowed unverified outbound sessions, we'd be able to use them for TALKREQ/RESP at least.

Potentially, though at least in a Portal Network context, we have no inbuilt way to obtain a full `Multiaddr` (including the public key).  We do have one subprotocol that relies entirely upon `TALKREQ` and drops the corresponding TALKRESPs and it could be useful to allow that subprotocol to work based just upon the INodeAddress.

**Comment by acolytec3** [comment]
And now that I think about it, another way to approach this might be to have the sessionService "established" event include the observed socket address along with the ENR as well as a "verified" flag and then update the routing table to only if the verified is set.  Then, Portal Network can listen for that and update our own ENR cache (since we are already also maintaining our own routing tables) with the ENRs and multiaddrs and then we can call `TALKREQ/TALKRESP` with ENR or multiaddr.  Thoughts?

File: src/service/service.ts (hunk 2 above)
Comment (acolytec3) [reply src/service/service.ts]: Disregard.  I was looking at the wrong method.  This feedback should now be addressed.

**Comment by acolytec3** [comment]
Disregard my previous feedback, I think this is now ready for review again.

**Comment by acolytec3** [comment]
> Does it make sense to also also allow outbound unverified sessions? They seem less useful, but you could dial a peer via a full multiaddr, and currently, if they respond with 'unverified' ENR (wrong or no IP+port), the session is dropped. If we allowed unverified outbound sessions, we'd be able to use them for TALKREQ/RESP at least.
> 

I went ahead and added this in as I think I now can see where it can be useful down the line.  I can't think of a use case where we'll be able to make use of it within Portal Network in our current code base unless we set up an ENR/socket address cache since we wouldn't have any other way to know the correct socket address to send a TALKREQ to but as I noted above, we do have one use case where we might want to still send outbound messages to a node with an invalid ENR so this will be a useful stepping stone to that use case.

File: src/session/service.ts [hunk 4]
```diff
@@ -510,28 +507,35 @@ export class SessionService extends (EventEmitter as { new (): StrictEventEmitte
...
-      // Verify the ENR is valid
-      if (this.verifyEnr(enr, nodeAddr)) {
-        // Session is valid
-        // Notify the application
+      // Verify the ENR endpoint matches observed node address
+      const verified = this.verifyEnr(enr, nodeAddr);
+
+      // Drop session if invalid ENR and session service not configured to allow unverified sessions
+      if (!verified && !this.config.allowUnverifiedSessions) {
+        log("ENR contains invalid socket address. Dropping session with %o", nodeAddr);
```
Comment (wemeetagain) [inline src/session/service.ts:515]: I think there should be a `this.failSession(nodeAddr, RequestErrorType.InvalidRemoteENR, true)` after this log and the similar log below.

**Review by wemeetagain — CHANGES_REQUESTED** [review]
> And now that I think about it, another way to approach this might be to have the sessionService "established" event include the observed socket address along with the ENR as well as a "verified" flag and then update the routing table to only if the verified is set. Then, Portal Network can listen for that and update our own ENR cache (since we are already also maintaining our own routing tables) with the ENRs and multiaddrs and then we can call `TALKREQ/TALKRESP` with ENR or multiaddr. Thoughts?

I think thats a good idea. We can update the `established` event type to:
`established: (nodeAddr: INodeAddress, enr: ENR, connectionDirection: ConnectionDirection, verified: boolean) => void;`
and filter out unverified sessions as a first step in `onEstablished`.

This also wouldn't effect performance/behavior in the default configuration, since only verified sessions will emit `established` events.

File: src/session/service.ts (hunk 4 above)
Comment (acolytec3) [reply src/session/service.ts:515]: Added.

**Comment by acolytec3** [comment]
> I think thats a good idea. We can update the established event type to:
established: (nodeAddr: INodeAddress, enr: ENR, connectionDirection: ConnectionDirection, verified: boolean) => void;
and filter out unverified sessions as a first step in onEstablished.

I've made these updates.  Can we also bubble up this `established` event to the discv5 service layer?  I'd have to add a new event at that level.  Otherwise, I can hack around it with a little `;(discv5 as any).sessionService.on("established...` in our portalnetwork code if you don't want it exposed in the main API.

**Comment by wemeetagain** [comment]
> > I think thats a good idea. We can update the established event type to:
> > established: (nodeAddr: INodeAddress, enr: ENR, connectionDirection: ConnectionDirection, verified: boolean) => void;
> > and filter out unverified sessions as a first step in onEstablished.
> 
> I've made these updates. Can we also bubble up this `established` event to the discv5 service layer? I'd have to add a new event at that level. Otherwise, I can hack around it with a little `;(discv5 as any).sessionService.on("established...` in our portalnetwork code if you don't want it exposed in the main API.

Maybe we can just make the `sessionService` public. You could then subscribe to events there without having to cast to any. Make it up to consumers of the library to not meddle with the sessionService in sketchy ways.

**Review by wemeetagain — APPROVED** [review]
Awesome! this will be a nice feature
